### PR TITLE
Feature: Support memory qos with cgroups v2

### PR DESCRIPTION
**Pull request by borgerli**
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Implements [KEP - Support memory qos with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos)

#### Does this PR introduce a user-facing change?

```release-note
Introducing Memory QoS support with cgroups v2 (Alpha)
The MemoryQoS feature is now in Alpha. This allows kubelet running with cgroups v2 to set memory QoS at container, pod and QoS level to protect and guarantee better memory quality. This feature can be enabled through feature gate MemoryQoS.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP] : [Support memory qos with cgroups v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos)

#### Note
After discussion with KEP owner @xiaoxubeii, this PR leaves the implementation of  kube-reserved / system-reserved cgroups for future.  The effective min boundary of a cgroup is also limited by memory.min values of all ancestor cgroups, and memory.min of all ancestor cgroups need to be set accordingly. But for kube-reserved / system-reserved cgroups, the ancestor cgroups may not be managed by kubelet.

### CODE REVIEW

**Comment by borgerli** [comment]
The first commit of this PR,"add unified on cri to support cgroup v2", is being reviewed in another PR: https://github.com/kubernetes/kubernetes/pull/102578. 
I'll rebase after PR102578 merged.

**Review by xiaoxubeii — CHANGES_REQUESTED** [review]

**Comment by xiaoxubeii** [comment]
lgtm 
ping @bobbypage @mrunalp for further review.

**Comment by bobbypage** [comment]
Closely related change to CRI to support unified - https://github.com/kubernetes/kubernetes/pull/102578

**Review by xiaoxubeii — APPROVED** [review]

**Comment by xiaoxubeii** [comment]
> Closely related change to CRI to support unified - #102578
Yes, I think this pr has rebased for that.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (bobbypage) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:112]: is it desirable to use the the same `memoryThrottlingFactor` for both calculating `memory.high` when memoryLimit exists as well as when memoryLimit does not exist and using node allocatable memory? Perhaps we want to have two different throttling factors? 

/cc @mrunalp

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (bobbypage) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:109]: can we add a comment here to explain logic from KEP:

i.e.

> If container sets limits.memory, we set memory.high=pod.spec.containers[i].resources.limits[memory] * memory throttling factor for container level cgroup if memory.high>memory.min

> If container doesn't set limits.memory, we set memory.high=node allocatable memory * memory throttling factor for container level cgroup

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (bobbypage) [inline pkg/kubelet/cm/cgroup_manager_linux.go:441]: one thing that may become confusing is when `Unified` should be used as opposed to existing `Resources` config from runc libcontainer. As I understand, some of the existing properties like `Memory` (https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/opencontainers/runc/libcontainer/configs/cgroup_linux.go#L49) already map to existing cgroupv2 items like `memory.Max`. 

However for the rest of the cgroupv2 options that are not implicitly converted underneath in `Resources` struct, `Unified` should be used instead. Is my understanding here correct?

/cc @giuseppe

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (bobbypage) [inline pkg/kubelet/cm/cgroup_manager_linux.go:442]: does this need to be guarded with `IsCgroupV2` check?

File: pkg/kubelet/cm/helpers_linux_test.go
Comment (bobbypage) [inline pkg/kubelet/cm/helpers_linux_test.go:626]: does it make sense to add some unit tests for the configuration of the QoS cgroup classes / configuration of the container resource config?

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (mrunalp) [reply pkg/kubelet/cm/cgroup_manager_linux.go:441]: Yes, that's right.

File: pkg/kubelet/cm/node_container_manager_linux.go
Comment (borgerli) [reply pkg/kubelet/cm/node_container_manager_linux.go:144]: This is to enforce the memory qos for node according to kube-reserved or system-reserved setttings, implementing points 1 & 2 at https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos#node

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (borgerli) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:40]: Maybe we could make this configurable in beta based on feedback from community users?

File: pkg/kubelet/cm/qos_container_manager_linux.go
Comment (borgerli) [reply pkg/kubelet/cm/qos_container_manager_linux.go:336]: Yes, qosContainerManagerImpl is used only when --cgroups-per-qos is set to true.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (xiaoxubeii) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:112]: I think one throttling factor makes it easier for understanding. When container sets the `memory_limit`, throttling is from the container perspective. If `memory limit` is not set, then it is from the global node perspective.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:40]: I think making it configurable will help us test with different values more easily but I am fine either ways.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:112]: I think this is okay for now but can revisit in next iteration.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (borgerli) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go]: fixed

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (borgerli) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:114]: fixed

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (borgerli) [reply pkg/kubelet/cm/cgroup_manager_linux.go]: fixed

File: pkg/kubelet/cm/helpers_linux_test.go
Comment (borgerli) [reply pkg/kubelet/cm/helpers_linux_test.go:626]: fixed.
1. QoS cgroup: add new test file: pkg/kubelet/cm/qos_container_manager_linux_test.go
2. container: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go::TestGenerateContainerConfigWithMemoryQoSEnforced

File: pkg/kubelet/cm/cgroup_manager_linux.go
Comment (giuseppe) [reply pkg/kubelet/cm/cgroup_manager_linux.go:441]: ideally unified is used for all the resources when running on cgroup v2.  It doesn't make difference for the memory.max limit, but for e.g. for the cpu controller you can specify the correct setting without relying on the conversions performed by the OCI runtime

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (borgerli) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go]: Fixed. Now it  can be configured in KubeletConfiguration.

**Comment by liggitt** [comment]
routing to @smarterclayton for review of kubelet config change per https://github.com/kubernetes/kubernetes/blob/7e00f5d401df08ea4c0c8982594fedade3a1118b/OWNERS_ALIASES#L385-L387

File: pkg/kubelet/cm/node_container_manager_linux.go
Comment (bobbypage) [reply pkg/kubelet/cm/node_container_manager_linux.go:144]: I see, it would be great to note this as a comment here or perhaps link to the KEP as you did in your comment.


One additional question, is this code path (i.e. `enforceExistingCgroup()`) called specifically only for when `enforce-node-allocatable=kube-reserved` and `enforce-node-allocatable=system-reserved` are set?

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go
Comment (bobbypage) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:148]: nit: consider refactoring this to a table driven test. That way it will be more clear for two separate test cases (since you can give them an name) and avoid some of the duplication.

i.e. test case 1 -- validate that pod with requests & limits has memory.min == requests ; memory.high == limits * throttling factor
test case 2 -- validate that pod with requests & no limits has memory.min == requests; memory.high == memoryNodeAllocatable * throttling factor

File: pkg/kubelet/cm/helpers_linux_test.go
Comment (bobbypage) [reply pkg/kubelet/cm/helpers_linux_test.go:626]: Nice, thanks for adding these. Do we also have test for node container manager changes? i.e. as described here https://github.com/kubernetes/kubernetes/pull/102970/files#r661921645

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go
Comment (borgerli) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:148]: Yes for 
> One additional question, is this code path (i.e. enforceExistingCgroup()) called specifically only for when enforce-node-allocatable=kube-reserved and enforce-node-allocatable=system-reserved are set?

Refactored the test case accordingly, Thanks.

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (borgerli) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:41]: fixed.

File: pkg/kubelet/apis/config/types.go
Comment (borgerli) [reply pkg/kubelet/apis/config/types.go]: fixed

**Comment by borgerli** [comment]
The check of pull-kubernetes-e2e-gce-alpha-features always failed due to kube-scheduler not running with following errors:
## errors in kube-scheduler  log
I0702 09:43:59.647023       9 flags.go:59] FLAG: --write-config-to=""
profiles[0].pluginConfig[5].args.shape: Required value: at least one point must be specified

## errors in build log
I0702 11:17:01.493] scheduler            Unhealthy   Get "http://127.0.0.1:10251/healthz": dial tcp 127.0.0.1:10251: connect: connection refused   
I0702 11:17:01.493] etcd-1               Healthy     {"health":"true","reason":""}                                                                 
I0702 11:17:01.493] etcd-0               Healthy     {"health":"true","reason":""}                                                                 
I0702 11:17:01.494] controller-manager   Healthy     ok                                                                                            
I0702 11:17:01.494] [0;31mValidation returned one or more failed components. Cluster is probably broken.[0m

File: pkg/kubelet/cm/qos_container_manager_linux_test.go
Comment (bobbypage) [inline pkg/kubelet/cm/qos_container_manager_linux_test.go:150]: nit: leave comment of where magic 384Mi comes from (i.e. `128+256` from burstable-pod-1 + burstable-pod-2 memory requests).

File: pkg/kubelet/cm/qos_container_manager_linux_test.go
Comment (bobbypage) [inline pkg/kubelet/cm/qos_container_manager_linux_test.go:151]: nit: leave comment of where 128Mi comes from (i.e. `guaranteed-pod` memory requests).

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go
Comment (bobbypage) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:148]: thanks for and info, and thanks for refactoring test, looks much cleaner now.

**Comment by bobbypage** [comment]
Thanks for addressing comments, this looks good to me, thanks for all you efforts.

/lgtm

Would like others to also take a look and provide feedback.

**Comment by borgerli** [comment]
> Thanks for addressing comments, this looks good to me, thanks for all you efforts.
> 
> /lgtm
> 
> Would like others to also take a look and provide feedback.

Thank you so much for all the advice.

**Comment by xiaoxubeii** [comment]
/lgmt Thanks for contributions. Would like to also take a look? @smarterclayton @thockin @mrunalp

**Comment by borgerli** [comment]
@smarterclayton @thockin @mrunalp Could you please help review this PR? Thanks.

**Comment by borgerli** [comment]
> pull-kubernetes-e2e-gce-alpha-features

According to the job history: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-alpha-features, this check has been always failed since `Jul 02 10:15:06`.
I opened an issue for this: https://github.com/kubernetes/test-infra/issues/22796

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (thockin) [inline pkg/kubelet/apis/config/fuzzer/fuzzer.go:110]: why a constant and not random?

File: pkg/kubelet/apis/config/validation/validation.go
Comment (thockin) [inline pkg/kubelet/apis/config/validation/validation.go:216]: The usual pattern here is:

if gate enabled && nil {
  // error: required value
}
if !nil {
    // check values
}

File: pkg/kubelet/apis/config/validation/validation.go
Comment (thockin) [inline pkg/kubelet/apis/config/validation/validation.go:218]: use the API-style name `memoryThrottlingFactor` (lower-0case first char)

Comment (thockin) [reply pkg/kubelet/apis/config/validation/validation.go:218]: s/smaller/less

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (xiaoxubeii) [reply pkg/kubelet/apis/config/fuzzer/fuzzer.go:110]: https://github.com/kubernetes/enhancements/blame/master/keps/sig-node/2570-memory-qos/README.md#L95

> e.g. if a container has `requests.memory=50, limits.memory=100`, and we have a throttling factor of .8, `memory.high` would be 80. if a container had no memory limit specified, we substitute `limits.memory` for `node allocatable memory` and apply the throttling factor of .8 to that value.


According to the kep definition, the default value is 0.8. This is based on practical experience.

Comment (thockin) [reply pkg/kubelet/apis/config/fuzzer/fuzzer.go:110]: The point of a fuzzer is to try unexpected values.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (xiaoxubeii) [reply pkg/kubelet/apis/config/validation/validation.go]: Updated.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (xiaoxubeii) [reply pkg/kubelet/apis/config/validation/validation.go]: Updated.

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (xiaoxubeii) [reply pkg/kubelet/apis/config/fuzzer/fuzzer.go]: Thanks for reminding, have updated.

File: pkg/kubelet/apis/config/fuzzer/fuzzer.go
Comment (thockin) [inline pkg/kubelet/apis/config/fuzzer/fuzzer.go:111]: LOL>  I made you fix it but when I scroll up, this WHOLE FILE is wrong.  Sorry.

**Review by thockin — COMMENTED** [review]
API LGTM

**Comment by mrunalp** [comment]
/approve
for Node. Thanks!

**Review by ehashman — COMMENTED** [review]
Read through the implementation and SGTM.

Graduation criteria for alpha:
- [x] cgroup_v2 is in Alpha
- [x] Memory QoS is implemented for new feature gate (`MemoryQoS`)
- [ ] Memory QoS is covered by proper tests: We have unit tests but it does not appear there are any e2es?
- [x] Memory QoS supports containerd, cri-o

Given missing e2es I am not sure if this is good to go. Let's at least run this through alpha features and cgroupsv2 coverage.

**Comment by xiaoxubeii** [comment]
> Read through the implementation and SGTM.
> 
> Graduation criteria for alpha:
> 
> * [x]  cgroup_v2 is in Alpha
> * [x]  Memory QoS is implemented for new feature gate (`MemoryQoS`)
> * [ ]  Memory QoS is covered by proper tests: We have unit tests but it does not appear there are any e2es?
> * [x]  Memory QoS supports containerd, cri-o
> 
> Given missing e2es I am not sure if this is good to go. Let's at least run this through alpha features and cgroupsv2 coverage.

Can we use cgv2  e2e in alpha first?

**Comment by borgerli** [comment]
The check of `pull-kubernetes-node-kubelet-serial-crio-cgroupv2` always failed due to issue #103579.

**Comment by thockin** [comment]
/approve 

based on @bobbypage and @mrunalp and @ehashman reviews.

**Comment by ehashman** [comment]
> @ehashman Can we use cgv2 e2e in alpha first? It will take some time to submit e2e PR, which we can make up later.

Given that pull-kubernetes-e2e-gce-alpha-features is passing there is at least _some_ coverage for the feature flag being enabled, but since none of the logic of this PR is triggered without the feature flag enabled *and* cgroupsv2, IMO this needs a test job added. Since none of our available jobs are able to check this right now, I'm hesitant to merge without exercising these code paths in CI. This PR should have had some test-infra changes submitted simultaneously so that we can trigger such a job on a PR. See e.g. https://github.com/kubernetes/test-infra/blob/58ae43ff3e79c2958827883f5f5b0164084bfb23/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml#L377-L415

If we can't get something merged in test-infra by EOD to test this I'll defer to Node leads to make the call since this is alpha, but I really think we need to add a CI job. This hasn't actually run through CI so I don't have confidence that it works.

**Comment by ehashman** [comment]
/hold 

to resolve discussion

**Comment by ehashman** [comment]
Job passed! :partying_face: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102970/pull-kubernetes-node-memoryqos-cgrpv2/1413181901090328576/

/hold cancel
/approve

**Pull request merged**
